krun/util.py

```diff
@@ -76,6 +104,89 @@ def run_shell_cmd_list(cmds, failure_fatal=True, extra_env=None):
         _, _, rv = run_shell_cmd(cmd, extra_env=extra_env)
         assert rv == 0
 
+def print_stderr_linewise(info):
+    stderr_partial_line = []
+    while True:
+        d = yield
+        # Take what we just read, and any partial line we had from
+        # a previous read, and see if we can make full lines.
+        # If so, we can print them, otherwise we keep them for
+        # the next time around.
+        startindex = 0
+        while True:
+            try:
+                nl = d.index("\n", startindex)
+            except ValueError:
+                stderr_partial_line.append(d[startindex:])
+                break  # no newlines
+            emit = d[startindex:nl]
+            info("stderr: " + "".join(stderr_partial_line) + emit)
+            stderr_partial_line = []
+            startindex = nl + 1
+
+def read_popen_output_carefully(process, platform, print_stderr=True):
+    """ helper function: given a process, read new data whenever it is
+    available, to make sure that the process is never blocked when trying to
+    write to stdout/stderr. """
+    from krun.platform import detect_platform
```

Comment:
[I know this code is sort-of copy and pasted, but sometimes a fresh look at things raises new questions!]

I think the `import` statement should be inside the `if` since that's the only place that `detect_platform` is called. Also... I wonder why this function even *does* platform detection itself. Shouldn't users be forced to do this?